### PR TITLE
Proof of concept to release snapshots to sonatype

### PR DESCRIPTION
**Pull request by georgeblahblah**
Co-authored-by: David Lawes <david.lawes@guardian.co.uk>

## What does this change?

This PR adds a proof of concept for releasing snapshots to sonatype via github actions. This PR makes a couple of changes:
- We define a new github workflow: the workflow is run when a release is published and only when the release is a "prerelease". The workflow references repository secrets (replicating what we expect local users to define) and triggers sbt to release a snapshot
- We started mapping out the release steps we'd need for the snapshot release. This is a work in progress, and there could be some refactoring after we've worked out what the non-snapshot release process could be.
- We had to bump the version of the sbt-pgp plugin to enable signing to take place successfully.

## How was this tested?

During development the github workflow was triggered on push events. For the final configuration of the work flow we can see the [action](https://github.com/guardian/content-api-scala-client/actions/runs/3968609802) was successfully completed:
- workflow completed successfully
- snapshot releases could be found in maven (content api client 2.12, content api client 2.13, content api client default 2.12, content api client 2.13 https://oss.sonatype.org/content/repositories/snapshots/com/gu/)
Screenshot 2023-01-20 at 15 22 03
Screenshot 2023-01-20 at 15 22 43

## What next?

After this change is deployed snapshot releases _could_ be made via the github UI by creating a new prerelease. Creating the new prerelease should trigger the workflow (including parsing the version correctly from the release name). This last part of the flow hasn't been tested yet (it can't be until this change has been merged?). After this change has been merged we'd want to:
- make a prerelease
- confirm the workflow is triggered
- confirm the workflow parses the expected version number
- confirm the snapshot releases are created

If we're happy with the prerelease proof of concept we could then:
- develop an equivalent workflow for releasing new non-snapshot releases via the github ui
- update the readme/docs describing how to release (including naming conventions for each release)

### CODE REVIEW

File: .github/workflows/release-snapshot.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+name: Release Snapshot to Sonatype
+
+on:
+  release:
+    types: [published]
+
+jobs:
+  release_snapshot_sonatype:
+    if: "github.event.release.prerelease"
```
Comment (DavidLawes) [inline .github/workflows/release-snapshot.yml:9]: In the github UI, when creating a snapshot release, we'll expect that the release to be marked as one
Screenshot 2023-01-20 at 15 04 39

File: .github/workflows/release-snapshot.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-java@v3
+        with:
+          distribution: corretto
+          java-version: 8
+          cache: sbt
+      - name: Release Snapshot to Sonatype
+        run: |
+          VERSION=$(git describe --tags | cut -f2 -d"@")
```
Comment (DavidLawes) [inline .github/workflows/release-snapshot.yml:20]: We're expecting a snapshot release to named with specific syntax:
`X.X.X-SNAPSHOT`

Note the missing `v` from the beginning of the release title, and the mandatory `-SNAPSHOT` at the end. Probably there'd need to be extra instructions in the readme about how to handle this

File: .github/workflows/release-snapshot.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+          echo $PGP_SECRET | base64 --decode | gpg --batch --import
+          export GPG_TTY=$(tty)
+          echo "Releasing version $VERSION Sonatype as snapshot"
+          sbt "clean" "project clientCI" "set isSnapshot := true" "release release-version $VERSION with-defaults"
+        env:
+          PGP_SECRET: ${{ secrets.PGP_SECRET }}
```
Comment (DavidLawes) [inline .github/workflows/release-snapshot.yml:26]: These secrets have been added to the repository. There is no org-wide pgp key or sonatype account, so I've had to use my own personal creds. I don't _think_ there's a better alternative (apart from replacing these with keys from someone in the CAPI team?)

File: project/plugins.sbt [hunk 4]
```diff
@@ -1,4 +1,4 @@
-addSbtPlugin("com.jsuereth" % "sbt-pgp" % "2.0.1")
+addSbtPlugin("com.github.sbt" % "sbt-pgp" % "2.1.2")
```
Comment (DavidLawes) [inline project/plugins.sbt:1]: Before bumping the library I kept seeing a `file not found` error when attempting to sign artefacts. We can see from [here](https://github.com/sbt/sbt-pgp/releases) that a min of 2.1.1 might be necessary for enabling sbt-pgp to run successfully in CI?

Given it's not a major change, I'd hope there won't be breaking changes, but probably we'd want to test the existing (manual) release process isn't impacted by this.

File: build.sbt [hunk 5]
```diff
@@ -123,6 +150,27 @@ lazy val checkReleaseType: ReleaseStep = ReleaseStep({ st: State =>
   st
 })
 
+lazy val releaseCIProcessSteps: Seq[ReleaseStep] = {
+  val commonSteps = Seq(
+    // seems like we don't need this next step for the CI release process, we already know we're doing a snapshot release
```
Comment (DavidLawes) [inline build.sbt:155]: Some of these comments will become superfluous, but I've retained them for now as I think it'll help remind us (and work out what's needed) when it comes to defining the release process (non-snapshot releases).

File: .github/workflows/release-snapshot.yml (hunk 1 above)
Comment (fredex42) [reply .github/workflows/release-snapshot.yml:9]: 👍

File: .github/workflows/release-snapshot.yml (hunk 2 above)
Comment (fredex42) [reply .github/workflows/release-snapshot.yml:20]: Any chance that we could tolerate the "v" on the start by stripping it off here? Just thinking that it's likely to be an obscure failure

File: build.sbt [hunk 6]
```diff
@@ -33,6 +33,14 @@ lazy val root = (project in file("."))
...
   )
 
+lazy val clientCI = (project in file("client/target/client_ci"))
+  .settings(commonSettings, clientSettings, publishCISettings)
+  .settings(
+    Compile / scalaSource := baseDirectory.value / "client" / "src" / "main" / "scala",
+    releaseVersionFile := baseDirectory.value / ".." / "version.sbt",
+  )
+  .enablePlugins(BuildInfoPlugin)
+
```
Comment (fredex42) [inline build.sbt:43]: now it's proven, would be nice to remove the superfluous project

**Review by fredex42 — CHANGES_REQUESTED** [review]
Hi guys,

Apologies it's taken ages to get back.  Really like it, just a couple of requests:
- is it possible to automatically handle the case where a `v` gets put into the tag by accident? (i.e. strip it off the start)
- would be nice to remove the superfluous `clientCI` project now that it's proven, unless you think that there would be risk associated with this?

**Review by JustinPinner — CHANGES_REQUESTED** [review]
Can we also update the release docs in this PR before it gets merged? Thanks! See https://github.com/guardian/content-api-scala-client/blob/main/docs/release.md

File: build.sbt (hunk 6 above)
Comment (DavidLawes) [reply build.sbt]: We tried removing the project but actually there are some underlying processes/steps that aren't quite compatible (yet, or without further work on our part).

Would the intention be to only support CI releases of the project, or would you always want to allow both mechanisms for release (via CI and manually)?

File: .github/workflows/release-snapshot.yml [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        with:
+          distribution: corretto
+          java-version: 8
+          cache: sbt
+      - name: Release Snapshot to Sonatype
+        run: |
+          VERSION=$(git describe --tags | cut -f2 -d"@")
+          if [[ ${VERSION:0:1} == "v" ]] ; then
+          VERSION=${VERSION:1}
+          fi
```
Comment (DavidLawes) [inline .github/workflows/release-snapshot.yml:23]: Here we strip out a leading `v` if it's been provided

File: .github/workflows/release-snapshot.yml [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+          if [[ ${VERSION: -9} != "-SNAPSHOT" ]] ; then
+            echo "Version must end in -SNAPSHOT. Adding -SNAPSHOT suffix"
+            VERSION="$VERSION-SNAPSHOT"
+          fi
```
Comment (DavidLawes) [inline .github/workflows/release-snapshot.yml:27]: If the pre-release doesn't have a `-SNAPSHOT` suffix then we add it here

File: .github/workflows/release-snapshot.yml [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+          echo $PGP_SECRET | base64 --decode | gpg --batch --import
+          export GPG_TTY=$(tty)
+          echo "Releasing version $VERSION Sonatype as snapshot"
+          yes | sbt -DRELEASE_TYPE=snapshot "clean" "project client" "release release-version $VERSION with-defaults"
```
Comment (DavidLawes) [inline .github/workflows/release-snapshot.yml:31]: This pipes the output of `yes` into the sbt commands - this means we can integrate with the existing build projects (which require manual input).

Note that we also need to define the release type as a system prop to integrate with the current build steps too.

File: build.sbt (hunk 6 above)
Comment (DavidLawes) [reply build.sbt]: Actually, not to worry - we have a solution that integrates with the existing project now :)

File: .github/workflows/release-snapshot.yml (hunk 9 above)
Comment (DavidLawes) [reply .github/workflows/release-snapshot.yml:31]: I think this is only releasing for:

content-api-client_2.12
content-api-client_2.13

Maybe we'll need to run `release-cross` to release the default client at the same time?

Comment (georgeblahblah) [reply .github/workflows/release-snapshot.yml]: We created a CAPI key and added it to the repository secrets. We also removed `"project client"` from the `sbt` command, so the GitHub Action now releases both `content-api-client-default` and `content-api-client`

**Review by fredex42 — APPROVED** [review]

**Review by JustinPinner — APPROVED** [review]
Looks good to me 👍

**Pull request merged**